File: src/stack/bracketValidation.ts
import { Stack } from "./linkedListStack"

interface IBracketValidation {
  stack: typeof Stack
  opening: string[]
  closing: string[]
  checkForValidation: (input: string) => boolean
  isOpening: (input: string) => boolean
  isClosing: (input: string) => boolean
  isPair: (x: string, y: string) => boolean
}

const BracketValidation = function (this: IBracketValidation) {
  this.stack = new Stack<string>()
  this.opening = ["{", "(", "[", "<"]
  this.closing = ["}", ")", "]", ">"]

  /**
   * this method checks if the paranthesis inside the string are balanced or not
   * @param {string} input string with paranthesis of type:any
   * @returns boolean true:balance paranthesis, false: non-balance paranthesis
   */
  this.checkForValidation = function (input: string) {
    for (const char of input) {
      if (this.isOpening(char)) {
        this.stack.push(char)
      } else if (this.isClosing(char)) {
        if (
          this.stack.isEmpty() ||
          !this.isPair(this.stack.peek().data, char)
        ) {
          return false
        } else {
          this.stack.pop()
        }
      }
    }
    return this.stack.isEmpty()
  }
  /**
   * checks if the character-sympol is an opening paranthessis
   * @param {string} char char of a string
   * @returns boolean true: opening paranthesis, false: non-opening paranthesis
   */
  this.isOpening = function (char: string) {
    return this.opening.indexOf(char) >= 0
  }
  /**
   * checks if the character-sympol is a closing paranthessis
   * @param {string} char char of a string
   * @returns boolean true: closing paranthesis, false: non-closing paranthesis
   */
  this.isClosing = function (char: string) {
    return this.closing.indexOf(char) >= 0
  }
  /**
   * checks if the top of the stack pairs with the closing tag found in the string
   * @param {string} opening paranthesis from the top of the stack
   * @param {string} closing paranthesis
   * @returns boolean true: opening, closing are pairs.
   */
  this.isPair = function (opening, closing) {
    return this.closing.indexOf(closing) === this.opening.indexOf(opening)
  }
}

export { BracketValidation }

File: src/stack/bracketValidation.c
/*
  C Program to check for balanced parentheses in an expression using stack.
  Given an expression as string comprising of opening and closing characters
  of parentheses - (), curly braces - {} and square brackets - [], we need to
  check whether symbols are balanced or not.
*/
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#define MAX 100

struct Node
{
  int data;
  struct Node *next;
};
typedef struct Node node;

node *top = NULL;

// PUSH()
void Push(char c)
{
  node *temp = (node *)malloc(sizeof(node));
  temp->data = c;
  temp->next = top;
  top = temp;
}

// POP()
void Pop()
{
  node *temp;
  if (top == NULL)
  {
    return;
  }
  temp = top;
  top = top->next;
  free(temp);
}
// Function to check whether two characters are opening
// and closing of same type.
int isMatching(char c)
{
  if (top->data == '(' && c == ')')
  {
    return 1;
  }
  else if (top->data == '[' && c == ']')
  {
    return 1;
  }
  else if (top->data == '{' && c == '}')
  {
    return 1;
  }

  else
    return 0;
}

void isBalanced(char exp[], unsigned long len)
{
  for (int i = 0; i < len; i++)
  {
    if ((exp[i] == '(') || (exp[i] == '{') || (exp[i] == '['))
    {
      Push(exp[i]);
    }
    else if ((exp[i] == ')') || (exp[i] == ']') || (exp[i] == '}'))
    {
      if (top == NULL)
      {
        return;
      }
      if (isMatching(exp[i]))
      {
        Pop();
      }
      else
        return;
    }
  }
  return;
}

int main(void)
{
  char exp[MAX];
  unsigned long len;
  printf("Enter expression: ");
  scanf("%s", exp); // input expression from console
  len = strlen(exp);
  isBalanced(exp, len);
  if (top == NULL)
  {
    printf("Balanced\n");
  }
  else
    printf("Not Balanced\n");

  return 0;
}

/*
src/stack> gcc ./bracketValidation.c
src/stack> a.out

your valid expression => output will be balanced
your non-valid expression => output will be Not Balanced
*/
